clients/notification-pubsub/src/main/java/org/eclipse/hono/client/notification/pubsub/PubSubBasedNotificationReceiver.java

```diff
@@ -0,0 +1,127 @@
+        this.receiver = (pubsubMessage, ackReplyConsumer) -> {
+            handleMessage(pubsubMessage);
+            ackReplyConsumer.ack();
+        };
+    }
+
+    /**
+     * Creates a client for consuming notifications. To be used for unittests.
+     *
+     * @param factory The factory to use for creating Pub/Sub subscribers.
+     * @param receiver The receiver to be called when messages are received.
+     * @throws NullPointerException If any of the parameters are {@code null}.
+     */
+    protected PubSubBasedNotificationReceiver(final PubSubSubscriberFactory factory, final MessageReceiver receiver) {
+        this.factory = Objects.requireNonNull(factory);
+        this.receiver = Objects.requireNonNull(receiver);
+    }
+
+    @Override
+    public <T extends AbstractNotification> void registerConsumer(final NotificationType<T> notificationType,
+            final Handler<T> consumer) {
+        if (notificationTypes.contains(notificationType.getAddress())) {
+            log.debug("Notification receiver {} is already registered", notificationType.getAddress());
+            return;
+        }
+        notificationTypes.add(notificationType.getAddress());
+
+        handlerPerType.put(notificationType.getClazz(), consumer);
+        final String subscriptionId = PubSubMessageHelper.getTopicName(TOPIC_ENDPOINT, notificationType.getAddress());
+        factory.getOrCreateSubscriber(subscriptionId, receiver)
+                .subscribe(true)
+                .onFailure(t -> {
+                    log.error("Error subscribing for notification {}", notificationType.getAddress(), t);
+                    throw new IllegalStateException("Error, can not subscribe for notification", t);
+                });
+    }
+
+    @Override
+    public Future<Void> start() {
+        if (lifecycleStatus.isStarting()) {
+            log.debug("Pub/Sub based notification receiver already started");
+            return Future.succeededFuture();
+        } else if (!lifecycleStatus.setStarting()) {
+            return Future.failedFuture(new IllegalStateException("Consumer is already started/stopping"));
+        }
+        return Future.succeededFuture();
+    }
+
+    @Override
+    public Future<Void> stop() {
+        return factory.closeAllSubscribers();
+    }
+
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    private void handleMessage(final PubsubMessage message) {
+        final Buffer buffer = Buffer.buffer(PubSubMessageHelper.getPayload(message));
+        final JsonObject jsonObject = buffer.toJsonObject();
+
+        if (log.isTraceEnabled()) {
+            log.trace("received notification: {}{}", System.lineSeparator(), jsonObject.encodePrettily());
+        }
+
+        final AbstractNotification notification = jsonObject.mapTo(AbstractNotification.class);
+        final Handler handler = handlerPerType.get(notification.getClass());
+        if (handler != null) {
+            handler.handle(notification);
+        }
```

Comment:
I think this can't be moved to the top because we only get the handler by doing all the steps before..
We are converting the data of a PubsubMessage into a Buffer object, this buffer object is converted into a jsonObject and this jsonObject is mapped to a notification. Based on this notification we can then check if a handler exists .. 

I renamed the variable to _'json'_ so that it is more understandable

Reply:
Oh, my bad, you're right, I didn't see the JSON being mapped to the class. In that case, however, you might also want to consider explicitly catching the runtime exceptions that are potentially thrown by Buffer.toJsonObject and by JsonObject.mapTo() ...